### ISSUE TITLE
Is it safe to run two InProcessConnections simultaneously on the same database?

### ISSUE THREAD

**Issue by 3noch** [agentm/project-m36#102]
Most commonly, I'd want to run `tutd` on a database while a server is running.

**Comment by agentm** [comment]
```InProcessConnection```s can be used to create any number of in-process databases and thus the ```Connection``` is not shared within a single process (no ```IORef```s). If you need multiple, remote clients then the only option is the ```RemoteProcessConnection``` which sets up a proper server with sockets, etc.

Use the ```InProcessConnection``` for SQLite-like capabilities.

(Behind the scenes, the ```RemoteProcessConnection``` is a wrapper implemented over the ```InProcessConnection```, so the features not related to the server are the same.)

**Comment by 3noch** [comment]
Ok, so I'm not 100% clear if it's safe for me to run, for example, a web app using `my.db` (connecting with `InProcessConnection`) and then via the terminal use `tutd -d my.db` while the application is running.

**Comment by agentm** [comment]
Oh, I see. You mean you wish to share the database directory among different servers. My plan was for this to work because all file writes are atomic, but I suspect I have a bug whereby the transaction graph is not reloaded when the file changes so that's the issue which prevents this from working properly. Currently, multiple servers to the same database directory would likely overwrite each other's changes to the graph, but the transactions would be persisted.

I know that sqlite has struggled with similar synchronization issues, so I need to add some file synchronization.

On the other hand, if you don't commit any transactions, read-only access should be completely safe.

In other words, you are treading into unknown territory.

**Comment by 3noch** [comment]
Ok thanks for the clarity. This would be a "nice-to-have" but certainly not essential. SQLite is optimized for this use case specifically (because it has no server and never will). PG/MySQL always connect to a server. So having *both* would be unique indeed.

**Comment by 3noch** [comment]
I've concluded rather quickly that it's not safe to have two servers running on the same database directory. I had an `InProcessConnection` in `ghci` and also had `tutd` open. My commits in `tutd` were lost.

**Comment by agentm** [comment]
Thanks for trying it out. This could be made to work with advisory file locking around a single file- the transaction graph log- but there could be some new errors generated due to graph contention. I will need to think more about how best to resolve such conflicts. 

Currently, each backend assumes that it can store a definitive transaction graph in memory and just flushes to disk on commit. That's the source of the lost transactions.

**Comment by 3noch** [comment]
At the very least, would it be possible for a second client to recognize that another client has primary ownership of the transaction log and reject all commits?

**Comment by agentm** [comment]
Sorry, that's not possible. Each ```InProcessConnection``` is a separate database. To share databases, use ```RemoteProcessConnection``` which uses STM to synchronize access. Maybe the naming scheme could be improved to make this clearer, though we don't offer a specific connection type for multi-threaded in-process access.

**Comment by agentm** [comment]
Sorry, I misunderstood your question in the context. I would like to fix the feature so that it behaves like SQLite using locking. 

Or perhaps I don't quite understand your use case- are you intentionally avoiding the ```RemoteProcessConnection```?

**Comment by agentm** [comment]
Thinking about this some more, perhaps I understand your use case, but do correct me if I am wrong.

You are looking to share the database amongst multiple temporary connection, correct? So, you have a web server which opens temporary connections to the database for each web request and a temporary interactive ```tutd``` session. In this case, we don't yet operate like SQLite and you will want to run a permanent ```project36-server``` instance to serve the database.

**Comment by agentm** [comment]
With the current implementation, you can get temporary, read-only snapshots with multiple clients reading the database directory directly- that's safe- but you can't commit: some commits will be lost. The read-only-ness is not enforced, which is what you are requesting. But I think we can do better and offer an equivalent to SQLite with proper file locking.

**Comment by 3noch** [comment]
Thanks here's my use case: I have a server and 99.9% of the time it will be the only thing accessing the database, so I feel inclined to use `InProcessConnection` because I don't really need multiple clients. However, sometimes I may want to debug the server with a `tutd` session just to see what's going on. I don't want to fear that I may ruin live production transactions with my `tutd` session which is why I want at least the safety on not clobbering anything.

**Comment by agentm** [comment]
I see, so a read-only connection would be of general interest then.

I suspect you are hitting the same wall that other developers hit when they have to choose between SQLite or a database server. Project:M36 simply happens to support both connection types, making the decision more difficult. I will have to think about how to remedy this, at least with some recommendations in the documentation. Certainly, getting on-par feature-wise with SQLite would make the distinction less important. 

Keep in mind that, unlike SQLite/PostgreSQL, Project:M36 offers a direct connection type promotion path which seems to be a big advantage. If the ```InProcessConnection``` is no longer sufficient, just point a ```project-m36-server``` at the same database directory and you are ready-to-go.

**Comment by 3noch** [comment]
> just point a project-m36-server at the same database directory and you are ready-to-go.

Yes that's a very nice feature. My only "concern" is that the in-process connection is too weak/dangerous to be useful for anything but toy applications. Any production scenario will need *some* sort of concurrent access, even if minimally (as in my case). To make in-process databases useful in production scenarios I think that it should at least support some safety measures to make sure it doesn't break things. Enforced read-only access for secondary connections would be workable. At this point I'm worried that I will break something in `tutd` so while using it I'm inclined to either take down the server (which is not generally viable) or just use remote connections. However, it seems that the value of in-process connections is too quickly lost.

**Comment by agentm** [comment]
Since 771ba49eff30e421b65f2aa80fd047f276105ef0, it is safe to share database directories amongst multiple processes (server or in-process instances).

[Replication Documentation](https://github.com/agentm/project-m36/blob/master/docs/replication.markdown)

There is still some low-hanging, performance-improving fruit to be implemented, but I'll open separate tickets for that.
